ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/ScalablePushRegistry.java

```diff
@@ -167,33 +191,40 @@ public synchronized boolean isClosed() {
    * Registers a ProcessingQueue with this scalable push registry so that it starts receiving
    * data as it streams in for that consumer.
    * @param processingQueue The queue to register
-   * @param expectingStartOfRegistryData Whether the request is expecting a new server without
-   *                                     rows received.
+   * @param catchupMetadata The catchup metadata including offsets to start at
    */
   public synchronized void register(
       final ProcessingQueue processingQueue,
-      final boolean expectingStartOfRegistryData
+      final Optional<CatchupMetadata> catchupMetadata
   ) {
     if (closed) {
       throw new IllegalStateException("Shouldn't register after closing");
     }
-    final LatestConsumer latestConsumer = this.latestConsumer.get();
-    if (latestConsumer != null && !latestConsumer.isClosed()) {
-      if (latestConsumer.getNumRowsReceived() > 0
-          && newNodeContinuityEnforced && expectingStartOfRegistryData) {
-        LOG.warn("Request missed data with new node added to the cluster, rows {}",
-            latestConsumer.getNumRowsReceived());
-        throw new KsqlException("New node missed data");
+    try {
+      if (catchupMetadata.isPresent()) {
+        if (catchupCounter.get() >=
```

Comment:
I think I left this comment elsewhere as well, but I think it's a bit misleading to use `AtomicInteger` here when everything is synchronized anyway. When I saw below us `getAndIncrement` my first thought was "hold up, there's probably a race with whatever is checking it's contents that might let us get two into this block", then I noticed it's synchronized so we don't actually need to worry.

I'm worried that future engineers will make false assumptions (e.g. "oh I don't need synchronized because it's atomic")

Reply:
Yeah, I agree.  The issue is that the decrement was happening outside of synchronized, which I considered a fine thing (maybe it was below the limit by the time the error was thrown).  In this case, I agree it's easier to just use synchronized.